v2/relay/relay.go

```diff
@@ -0,0 +1,483 @@
+func New(ctx context.Context, h host.Host, opts ...Option) (*Relay, error) {
+	ctx, cancel := context.WithCancel(ctx)
+
+	r := &Relay{
+		ctx:     ctx,
+		cancel:  cancel,
+		host:    h,
+		rc:      DefaultResources(),
+		acl:     nil,
+		rsvp:    make(map[peer.ID]time.Time),
+		refresh: make(map[peer.ID]time.Time),
+		conns:   make(map[peer.ID]int),
+	}
+
+	for _, opt := range opts {
+		err := opt(r)
+		if err != nil {
+			return nil, fmt.Errorf("error applying relay option: %w", err)
+		}
+	}
+
+	h.SetStreamHandler(ProtoIDv2Hop, r.handleStream)
+	h.Network().Notify(
+		&network.NotifyBundle{
+			DisconnectedF: r.disconnected,
+		})
+	go r.background()
+
+	return r, nil
+}
+
+func (r *Relay) Close() error {
+	select {
+	case <-r.ctx.Done():
+	default:
+		r.cancel()
+	}
+	return nil
+}
+
+func (r *Relay) handleStream(s network.Stream) {
+	select {
+	case <-r.ctx.Done():
+		s.Reset()
+		return
+	default:
+	}
+
+	s.SetReadDeadline(time.Now().Add(StreamTimeout))
+
+	log.Infof("new relay stream from: %s", s.Conn().RemotePeer())
+
+	rd := util.NewDelimitedReader(s, maxMessageSize)
+	defer rd.Close()
+
+	var msg pbv2.HopMessage
+
+	err := rd.ReadMsg(&msg)
+	if err != nil {
+		r.handleError(s, pbv2.Status_MALFORMED_MESSAGE)
+		return
+	}
+	// reset stream deadline as message has been read
+	s.SetReadDeadline(time.Time{})
+
+	switch msg.GetType() {
+	case pbv2.HopMessage_RESERVE:
+		r.handleReserve(s, &msg)
+
+	case pbv2.HopMessage_CONNECT:
+		r.handleConnect(s, &msg)
+
+	default:
+		r.handleError(s, pbv2.Status_MALFORMED_MESSAGE)
+	}
+}
+
+func (r *Relay) handleReserve(s network.Stream, msg *pbv2.HopMessage) {
+	defer s.Close()
+
+	p := s.Conn().RemotePeer()
+	a := s.Conn().RemoteMultiaddr()
+
+	if r.acl != nil && !r.acl.AllowReserve(p, a) {
+		log.Debugf("refusing relay reservation for %s; permission denied", p)
+		r.handleError(s, pbv2.Status_PERMISSION_DENIED)
+		return
+	}
+
+	r.mx.Lock()
+	now := time.Now()
+
+	refresh, exists := r.refresh[p]
+	if exists && refresh.After(now) {
+		// extend refresh time, peer is trying too fast
+		r.refresh[p] = refresh.Add(r.rc.ReservationRefreshTTL)
+		r.mx.Unlock()
+		log.Debugf("refusing relay reservation for %s; refreshing too fast", p)
+		r.handleError(s, pbv2.Status_RESERVATION_REFUSED)
+		return
+	}
+
+	active := len(r.rsvp)
+	if active >= r.rc.MaxReservations {
+		r.mx.Unlock()
+		log.Debugf("refusing relay reservation for %s; too many reservations", p)
+		r.handleError(s, pbv2.Status_RESOURCE_LIMIT_EXCEEDED)
+		return
+	}
+
+	r.rsvp[p] = now.Add(r.rc.ReservationTTL)
+	r.refresh[p] = now.Add(r.rc.ReservationRefreshTTL)
+	r.host.ConnManager().TagPeer(p, "relay-reservation", ReservationTagWeight)
+	r.mx.Unlock()
+
+	log.Debugf("reserving relay slot for %s", p)
+
+	err := r.writeResponse(s, pbv2.Status_OK, r.makeReservationMsg(p), r.makeLimitMsg(p))
+	if err != nil {
+		s.Reset()
+		log.Debugf("error writing reservation response; retracting reservation for %s", p)
+		r.mx.Lock()
+		delete(r.rsvp, p)
```

Comment:
Nvm, I see we do this in GC and it makes more sense to do it there in case the peer tries connecting to us again before the refresh interval has elapsed. This is smart.

Reply:
refresh logic has been removed.